sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Mode.scala

```diff
@@ -86,6 +75,66 @@ case class Mode(
+      a: ArrayType,
+      data: ArrayData): Seq[Any] = {
+    (0 until data.numElements()).map { i =>
+      data.get(i, a.elementType) match {
+        case k: UTF8String if a.elementType.isInstanceOf[StringType] &&
+          !a.elementType.asInstanceOf[StringType].supportsBinaryEquality
+          => CollationFactory.getCollationKey(k, a.elementType.asInstanceOf[StringType].collationId)
+        case k if a.elementType.isInstanceOf[StructType] =>
+          recursivelyGetBufferForStructType(
+            k.asInstanceOf[InternalRow].toSeq(a.elementType.asInstanceOf[StructType]).zip(
+              a.elementType.asInstanceOf[StructType].fields))
+        case k if a.elementType.isInstanceOf[ArrayType] =>
+          recursivelyGetBufferForArrayType(
+            a.elementType.asInstanceOf[ArrayType],
+            k.asInstanceOf[ArrayData])
+        case k => k
+      }
+    }
+  }
+
+  private def getBufferForComplexType(
+      buffer: OpenHashMap[AnyRef, Long],
+      d: DataType): Iterable[(AnyRef, Long)] = {
+   buffer.groupMapReduce {
+      case (key: InternalRow, _) if d.isInstanceOf[StructType] =>
+        recursivelyGetBufferForStructType(key.toSeq(d.asInstanceOf[StructType])
+          .zip(d.asInstanceOf[StructType].fields))
+      case (key: ArrayData, _) if d.isInstanceOf[ArrayType] =>
+        recursivelyGetBufferForArrayType(d.asInstanceOf[ArrayType], key)
+    }(x => x)((x, y) => (x._1, x._2 + y._2)).values
+  }
+
+  private def recursivelyGetBufferForStructType(
+      tuples: Seq[(Any, StructField)]): Seq[Any] = {
+    tuples.map {
+      case (k: String, field) if tuples.exists(f => f._2.dataType.isInstanceOf[StringType] &&
+        !f._2.dataType.asInstanceOf[StringType].supportsBinaryEquality &&
+        f._2.name == field.name) =>
+        CollationFactory.getCollationKey(UTF8String.fromString(k),
+          field.dataType.asInstanceOf[StringType].collationId)
+      case (k: UTF8String, field) if tuples.exists(f =>
+        f._2.dataType.isInstanceOf[StringType] &&
+        !f._2.dataType.asInstanceOf[StringType].supportsBinaryEquality &&
+        f._2.name == field.name) =>
+        CollationFactory.getCollationKey(k, field.dataType.asInstanceOf[StringType].collationId)
+      case (k, field: StructField) if tuples.exists(f => f._2.dataType.isInstanceOf[StructType] &&
+        f._2.name == field.name) =>
+        recursivelyGetBufferForStructType (
+        k.asInstanceOf[InternalRow].toSeq(field.dataType.asInstanceOf[StructType]).zip(
+          field.dataType.asInstanceOf[StructType].fields))
+      case (k, structField: StructField) if structField.dataType.isInstanceOf[ArrayType] &&
+        structField.dataType.asInstanceOf[ArrayType].existsRecursively(s =>
+          s.isInstanceOf[StringType] && !s.asInstanceOf[StringType].supportsBinaryEquality)
+      => recursivelyGetBufferForArrayType(
+          structField.dataType.asInstanceOf[ArrayType],
+          k.asInstanceOf[ArrayData])
```

Comment:
ditto for all the cases (and the ones above), use UnsafeRowUtils

Reply:
Done